packages/concerto-core/lib/decoratormanager.js

```diff
@@ -253,6 +253,178 @@ class DecoratorManager {
+                } else {
+                    typeCommandsMap.set(decoratorCommand.target.type, [{...decoratorCommand}]);
+                }
+            }
+
+            if(decoratorCommand.target.property) {
+                if (propertyCommandsMap.has(decoratorCommand.target.property)) {
+                    propertyCommandsMap.get(decoratorCommand.target.property).push(decoratorCommand);
+                } else {
+                    propertyCommandsMap.set(decoratorCommand.target.property, [{...decoratorCommand}]);
+                }
+            } else if(decoratorCommand.target.mapElement) {
+                if (mapElementCommandsMap.has(decoratorCommand.target.declaration)) {
+                    mapElementCommandsMap.get(decoratorCommand.target.declaration).push(decoratorCommand);
+                } else {
+                    mapElementCommandsMap.set(decoratorCommand.target.declaration, [{...decoratorCommand}]);
+                }
+            } else if(decoratorCommand.target.declaration) {
+                if (declarationCommandsMap.has(decoratorCommand.target.declaration)) {
+                    declarationCommandsMap.get(decoratorCommand.target.declaration).push(decoratorCommand);
+                } else {
+                    declarationCommandsMap.set(decoratorCommand.target.declaration, [{...decoratorCommand}]);
+                }
+            } else if(decoratorCommand.target.namespace) {
+                if (namespaceCommandsMap.has(decoratorCommand.target.namespace)) {
+                    namespaceCommandsMap.get(decoratorCommand.target.namespace).push(decoratorCommand);
+                } else {
+                    namespaceCommandsMap.set(decoratorCommand.target.namespace, [{...decoratorCommand}]);
+                }
+            }
+        });
+
+        return {
+            namespaceCommandsMap,
+            declarationCommandsMap,
+            propertyCommandsMap,
+            mapElementCommandsMap,
+            typeCommandsMap
+        };
+    }
+
+    /**
+     * Applies all the decorator commands from the DecoratorCommandSet
+     * to the ModelManager.
+     * @param {ModelManager} modelManager the input model manager
+     * @param {*} decoratorCommandSet the DecoratorCommandSet object
+     * @param {object} [options] - decorator models options
+     * @param {boolean} [options.validate] - validate that decorator command set is valid
+     * with respect to to decorator command set model
+     * @param {boolean} [options.validateCommands] - validate the decorator command set targets. Note that
+     * the validate option must also be true
+     * @param {boolean} [options.migrate] - migrate the decoratorCommandSet $class to match the dcs model version
+     * @returns {ModelManager} a new model manager with the decorations applied
+     */
+    static optimizedDecorateModels(modelManager, decoratorCommandSet, options) {
```

Comment:
Need to update update the method based on the specificity mentioned in the issue #856

Reply:
The name doesn't seem right, either

Reply:
fixed